crates/matrix-sdk/src/authentication/oidc/registration.rs

```diff
@@ -80,3 +89,375 @@ pub struct ClientRegistrationResponse {
+    /// variants.
+    pub fn new(non_localized: T, localized: impl IntoIterator<Item = (LanguageTag, T)>) -> Self {
+        Self { non_localized, localized: localized.into_iter().collect() }
+    }
+
+    /// Get the non-localized variant.
+    pub fn non_localized(&self) -> &T {
+        &self.non_localized
+    }
+
+    /// Get the variant corresponding to the given language, if it exists.
+    pub fn get(&self, language: Option<&LanguageTag>) -> Option<&T> {
+        match language {
+            Some(lang) => self.localized.get(lang),
+            None => Some(&self.non_localized),
+        }
+    }
+}
+
+impl<T> From<(T, HashMap<LanguageTag, T>)> for Localized<T> {
+    fn from(t: (T, HashMap<LanguageTag, T>)) -> Self {
+        Localized { non_localized: t.0, localized: t.1 }
+    }
+}
+
+#[derive(Serialize)]
+struct ClientMetadataSerializeHelper {
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    redirect_uris: Vec<Url>,
+    token_endpoint_auth_method: &'static str,
+    grant_types: BTreeSet<GrantType>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    response_types: Vec<ResponseType>,
+    application_type: ApplicationType,
+    #[serde(flatten)]
+    localized: ClientMetadataLocalizedFields,
+}
+
+impl From<ClientMetadata> for ClientMetadataSerializeHelper {
+    fn from(value: ClientMetadata) -> Self {
+        let ClientMetadata {
+            application_type,
+            grant_types: oauth_grant_types,
+            client_uri,
+            client_name,
+            logo_uri,
+            policy_uri,
+            tos_uri,
+        } = value;
+
+        let mut redirect_uris = None;
+        let mut response_types = None;
+        let mut grant_types = BTreeSet::new();
+
+        // Support for refresh tokens is mandatory.
+        grant_types.insert(GrantType::RefreshToken);
+
+        for oauth_grant_type in oauth_grant_types {
+            match oauth_grant_type {
+                OauthGrantType::AuthorizationCode { redirect_uris: uris } => {
+                    redirect_uris = Some(uris);
+                    response_types = Some(vec![ResponseType::Code]);
+                    grant_types.insert(GrantType::AuthorizationCode);
+                }
+                OauthGrantType::DeviceCode => {
+                    grant_types.insert(GrantType::DeviceCode);
+                }
+            }
+        }
+
+        ClientMetadataSerializeHelper {
+            redirect_uris: redirect_uris.unwrap_or_default(),
+            // We only support public clients.
+            token_endpoint_auth_method: "none",
+            grant_types,
+            response_types: response_types.unwrap_or_default(),
+            application_type,
+            localized: ClientMetadataLocalizedFields {
+                client_uri,
+                client_name,
+                logo_uri,
+                policy_uri,
+                tos_uri,
+            },
+        }
+    }
+}
+
+struct ClientMetadataLocalizedFields {
```

Comment:
Why do you need this type?

Reply:
Because (de)serialization of `Localized` values is complicated so we need to implement it manually. But we don't need to things manually for other fields so we just group the localizable fields in a struct and handle their serialization separately.

Reply:
Great, thanks. It may deserves a small documentation line to explain that maybe?

Reply:
I added docs to the type

Reply:
Thanks!